### DISCUSSION TITLE
Improve performance of Add multiple-to-multiple edges

### DISCUSSION THREAD

**Discussion by vitaliishandra** [JanusGraph/janusgraph#3158] [Q&A]
We need to create the Many-To-Many edges.
To test we used 760 Users and tried to create edges to 3975 content items.
It should create 3021000 edges in total.

According to the profile output this operation took 51 seconds to execure. The AddE step took 90% of the time.
![image](https://user-images.githubusercontent.com/56965286/181279217-23a81848-87df-4586-b0d5-99ecfe85ba03.png)

Example of query:
`g.V().hasLabel('User').as('u').V().hasLabel('Content').addE('Indirect').from('u').profile()`

We use cassandra for backend.

Are there any ideas how to improve edge creation (VM scaling, query rewrite, janusGraph configs)?

**Comment by vtslab** [comment]
You can find some reference numbers on write performance (of vertices) here:
[https://www.experoinc.com/post/janusgraph-nuts-and-bolts-part-1-write-performance ](https://www.experoinc.com/post/janusgraph-nuts-and-bolts-part-1-write-performance  )

First check your configs regarding properties containing the word "batch".
After that you can only parallelize the write operations:
- use multiple threads
- use mutiple janusgraph instances

Check if your Cassandra backend is holding up and add nodes if necessary.

Best wishes,    Marc

**Comment by li-boxuan** [comment]
You may want to create a composite index for a property called "type" which is basically a replication of your node's label. Then the query would be: `g.V().has("type, "User")` which would be faster. See https://github.com/JanusGraph/janusgraph/issues/283
